### ISSUE TITLE
[Premium] Pricing tier not allowed in existing resource group

### ISSUE THREAD

**Issue by wahyuen** [Azure/Azure-Functions#1194]
**Describe the bug**
When attempting to deploy Premium plan to a resource group that is existing AND the location of the premium plan differs from the location of the resource group, the deployment fails with the following error. This fails for both deployment manually through the Azure Portal and also via ARM templates.

`The pricing tier 'ElasticPremium' is not allowed in this resource group.`

**To Reproduce**
1. Go to an existing Resource Group located in Australia SouthEast
2. Create new Azure Functions Premium Plan on Australia East

**Expected behavior**
Azure Functions Premium Plan to be successfully created

**Additional context**
[https://github.com/MicrosoftDocs/azure-docs/issues/29134](https://github.com/MicrosoftDocs/azure-docs/issues/29134)

Existing Resource Group Location (**Australia SouthEast**)

deploy using your ARM template targetting Australia SouthEast - **Works**
deploy using your ARM template targetting Australia East - **Does not work**
deploy using the Azure Portal targetting Australia SouthEast - **Works**
deploy using the Azure Portal targetting Australia East - **Does not work**

**Comment by jeffhollan** [comment]
I suspect it has nothing to do with the resource group region but likely due to the fact this resource group may have other resources (specifically app service plans) in them. The Azure Functions Premium plan is only available in a sub-set of infrastructure in each region. Internally we call these "webspaces" or "stamps."  We use resource groups as one way to determine stamp - so if your resource group is already using a stamp that is non-premium enabled, it won’t work for now. As we light up premium in more stamps this should happen less, but even then combinations sometimes may give issue. Something we are tracking though. Does that explanation match your existing resource groups? I suspect a brand new resource group in  a different region than plan would still work. Let us know

**Comment by wahyuen** [comment]
@jeffhollan Yes, I believe that matches closely with out scenario. We ended up deleting our dev resource group and recreating to temporarily work around this issue.

Is there perhaps any recommended 'migration' path for production/customer resource groups where we wish to deploy the premium plan to? Is the only way forward, either creating a new resource group and having the premium plan run separately? We aren't in a scenario to simply 'blow away' our customer facing resource groups given they have existing data there!

**Comment by minhtrieu-aemo** [comment]
We trying to deploy new function apps that support VNET integration and also encountered this issue. 
I created a new resource group with several storage accounts tried deploying again and it failed with the same error "The pricing tier 'ElasticPremium' is not allowed in this resource group." Then i deleted the resource group and tried again and it worked.
Telling customers to create a new resource group and migrate everything across is a terrible solution to this problem

**Comment by alexkarcher-msft** [comment]
Hello,

We've built out a fix for this limitation, and it'll take about 8 weeks to get validated and globally deployed. This is a pretty meaty fix on the backend, and is going to clear up this class of problem across a few different services. Consider it a preview limitation until then 😉

Feel free to keep opening up other issues as you run into them! The feedback is really helpful to the public preview!

**Comment by minhtrieu-aemo** [comment]
Hi Guys,

How is the fix roll-out going? This issue totally kills our DevOps pipeline automation activities. I kick off an entire release and it fails 70% of the time. We delete the resource group and keep trying again until it works.

We are using "Australia East"

**Comment by Ztarbox** [comment]
Hello,

any updates on the roll-out status?
Still getting the error on my resource-group:

`PS C:\> az functionapp plan create -g mygroup -n myplan --sku EP1
Requested feature is not available in resource group launchpad. Please try using a different resource group or create a new one.`

**Comment by alexkarcher-msft** [comment]
We're targeting mid-September to enable the ability to create premium plans in resource groups with existing plans.

Just to confirm, creation works fine for you when you try to create a plan in a new, empty resource group?

**Comment by Ztarbox** [comment]
> We're targeting mid-September to enable the ability to create premium plans in resource groups with existing plans.
> 
> Just to confirm, creation works fine for you when you try to create a plan in a new, empty resource group?

I didn't understand that this restriction also exists.
For new resource groups I have no issue. 

Is there a way to switch from "normal" to "premium / elastic" in an existing plan? Or is this planned?

**Comment by naren-b** [comment]
> We're targeting mid-September to enable the ability to create premium plans in resource groups with existing plans.
> 
> Just to confirm, creation works fine for you when you try to create a plan in a new, empty resource group?

Thank you for the update. We have existing resource group and planning to upgrade some of the apps to premium app service plans. From last two months, we are waiting for this ability.

**Comment by nelsonmorais** [comment]
Hi @alexkarcher-msft, 
I still have issues when I try to create an Elastic Plan to host my Azure Functions on an existing Resource Group

You've mentioned that a fix should be in place in mid September, as we're now beyond that data can you confirm if the fix is already in place and just not working as expected, or can you provide a new data for the fix to be in place?

Here's the output I get from he ARM Template execution:
```
{
    "id": "/subscriptions/XXXXX/resourceGroups/XXXXX/providers/Microsoft.Resources/deployments/Microsoft.FunctionAppXXXXX/operations/XXXXX",
    "operationId": "XXXXX",
    "properties": {
        "provisioningOperation": "Create",
        "provisioningState": "Failed",
        "timestamp": "2019-09-18T10:31:16.7375426Z",
        "duration": "PT0.3545459S",
        "trackingId": "XXXXX",
        "statusCode": "BadRequest",
        "statusMessage": {
            "Code": "BadRequest",
            "Message": "Requested feature is not available in resource group XXXXX. Please try using a different resource group or create a new one.",
            "Target": null,
            "Details": [
                {
                    "Message": "Requested feature is not available in resource group XXXXX. Please try using a different resource group or create a new one."
                },
                {
                    "Code": "BadRequest"
                },
                {
                    "ErrorEntity": {
                        "ExtendedCode": "59324",
                        "MessageTemplate": "Requested feature is not available in resource group {0}. Please try using a different resource group or create a new one.",
                        "Parameters": [
                            "XXXXX"
                        ],
                        "Code": "BadRequest",
                        "Message": "Requested feature is not available in resource group XXXXX. Please try using a different resource group or create a new one."
                    }
                }
            ],
            "Innererror": null
        },
        "targetResource": {
            "id": "/subscriptions/XXXXX/resourceGroups/XXXXX/providers/Microsoft.Web/serverfarms/eu-tst-fnpremium",
            "resourceType": "Microsoft.Web/serverfarms",
            "resourceName": "eu-tst-fnpremium"
        }
    }
}
```

**Comment by alexkarcher-msft** [comment]
Hi @nelsonmorais 

The fix is not yet deployed. It's currently in testing, and we expect it to be completely rolled out around the beginning of November. We ran into some issues with the initial validation that proved to be more difficult to fix than we expected.

Alex

**Comment by nelsonmorais** [comment]
Hi @alexkarcher-msft,
Thanks for the quick response!
We're really looking forward to use this App Plan as we're being hit by the slow cold starts of the Consumption Plan and because we need the elasticity offered by this Elastic Premium Plan. 
As November is still a few weeks from now, is there any option to have 4 of our resource groups "migrated" so that we can start using the Elastic Premium Plan sooner than that?
We have quite a few resources and data on these RGs so deleting and recreating them is not an option we can take.
Thanks,
Nelson

**Comment by wahyuen** [comment]
Hi @alexkarcher-msft,

Just wanted to touch base to see how this issue was looking given Premium was just annouced as GA? Has this fix now been rolled out?

cheers
Wah

**Comment by alexkarcher-msft** [comment]
Hello,

We're now expected to start rolling out updated infrastructure that will support mixed resource groups by the beginning of November, with most sites migrated by mid 2020. Because of this delay we're pursuing a different fix that may be available sooner. I'll update this thread if we have an earlier ETA.

**Comment by barrettlee** [comment]
Hi @alexkarcher-msft, 

> We're now expected to start rolling out updated infrastructure that will support mixed resource groups by the beginning of November

it is already November 4th as you type post.   I see from the mixed tenses referencing past and present in your statement above that maybe you planned to complete a solution by early November but slipped. 
Is there at least a manual workaround yet?

**Comment by alexkarcher-msft** [comment]
Yes, we are delayed getting the infrastructure change in place and now expect to have most sites migrated by mid 2020. Because of this delay we're pursuing a different fix that may be available sooner. 

The current manual workaround is still to create a new resource group and deploy a Premium plan or Pv2 App Service plan before deploying any other resource.

**Comment by nelsonmorais** [comment]
Hi @alexkarcher-msft @jeffhollan ,

I really appreciate your transparency on this topic but I have to say that in terms of expectation management this is not going on the right way, let's see:

- Issue opened in April 2019
- On August we got a response that a solution should be available in September 2019
- In September we got a new response saying that a solution should be available in beginning of November
- In November we got yet another response saying that there's no solution yet and now the ETA is by mid 2020 or eventually something in between that we have no ideia when

Meanwhile we also need to do planning which becomes quite unpredictable given the situation described above.

The workarounds are really not feasible, we can't just delete a resource group and create a new one when there's multiple resources and data on a Production resource group where we want our functions to be in. Creating another resource group just to have the Premium Functions its also messy, and has an impact on the current CI/CD pipeline.

So, when can we expect a proper solution to the problem? And please don't give us another date that you cannot commit to, as that is worst than having a date that we don't like to ear from you.

Sorry for my strong opinion on the topic but I really feel it is about time to have a clear answer for this.

**Comment by jesserizzo** [comment]
@alexkarcher-msft @jeffhollan 
I'm running into this error a few times that are surprising given the above explanation. I've seen it once on a freshly created resource group with no other resources in it. I've also seen it in an existing resource group that had previously contained premium app service plans, where after deleting all the resources from the group an elastic premium plan failed to deploy. All these deploys are using ARM templates. A couple of questions:
Is it possible for an empty resource group to be assigned a "stamp" that is not premium capable? Is it possible for the assigned stamp to be changed on an existing empty resource group?
Is the fix for this still wait until mid 2020?

Thanks.

**Comment by jeffhollan** [comment]
Understand the frustration - let me try to add some context.  Realize this may be a little bit of a repeat but worth noting.  

🔔  TL;DR - working on it, definitely the type of problem that's hard to give a firm ETA, but the immediate solution is create the premium plan in a new resource group and you'll be unblocked.  The apps in the plan can be in any group, but the plan needs to be in a resource group that is locked to a supported stamp. 🔔 

Long answer: For a variety of historic reasons from the early days of Azure, when you create an App Service Plan in a resource group it gets assigned a specific pool of available VMs to run on.  Internally we call this a stamp.  The first app service plan in a resource group now locks any future app service plans in that resource group to the same stamp.  Both functions and app services share the same underlying compute to our users on demand.  

The downside is that as App Services continued to evolve it got bigger and better machines, or a variety of types of machines.  Sometimes these existed in different stamps.  The most notable example of this is we have some pools that run all Windows VMs, and some pools that run all Linux VMs.  So if you have a resource group with a Windows App Service plan, [you can't then deploy a Linux App Service plan in it](https://docs.microsoft.com/en-us/azure/app-service/containers/app-service-linux-intro#limitations) (because that stamp doesn't even have Linux VMs).  The other downside is specifically for Premium Functions, Premium plans run on the newest and largest VMs in our pools, but consumption plan functions don't.  So a large amount of these pools that have a consumption plan in them don't even have the VMs to run the premium plan.  When that's the case, you get this error.  If you put the Premium plan in a new resource group it will be assigned a stamp / pool that supports these new VMs.

As for the fix, we are rolling out a change so that these VM pools are much larger and much more diverse.  While that may sound super simple, there are hundreds of thousands of VMs that we are managing and updating and running customer workloads across the world.  Making these types of changes is done with a ton of care so it's not disruptive.  So it's caused the date to be a little more fluid than I'd prefer because updating hundreds of thousands of VM metadata across the globe safely with no down time is a massive undertaking.  It is in progress, a lot of the machines have already been updated in many regions, but still many thousands yet to update.  I know the current goal date but am hesitant to share as it has moved (and for good reasons) - but just know it's in progress, it's a massive undertaking, and hopefully in the coming months will be fully resolved.

**Comment by jab054** [comment]
I tried using the PowerShell commands:
Select-AzureRmSubscription -SubscriptionId [id]
 Set-AzureRmWebApp -Name "myFunction" -ResourceGroupName "myPremiumResourse" -AppServicePlan "myPremiumPlan"
and receive this error:
Set-AzWebApp: Operation returned an invalid status code 'Conflict'

I first tried to add a new function to my existing resource group as a Premium Plan, but that failed.  I then created a new resource group, added my new function and selected Premium.  That worked just fine.  Then I took one of my existing Consumption function apps and moved it into the Premium resource group.  Now that it is in the same group, I tried to change the plan using the commands above.

Any idea as to why this is failing?  I have several other Consumption apps that need to be moved and upgraded, and are quite complex.  Lots of settings, api keys, CORS etc.  It would be nice if we can migrate all of this over.

**Comment by alexkarcher-msft** [comment]
We are in the process of rolling out the change to enable co-mingled plan types. It's about 20% rolled out at the moment. There is no particular region that is fully enabled, but expect these issues to get less and less common over the next ~6 months of our deployment.

**Comment by KaiWalter** [comment]
@alexkarcher-msft is there a way for me to determine, whether the roll out touched my subscription or resource group?

**Comment by jab054** [comment]
I ended up backing up all my settings and keys.  Deleted my existing function apps and recreated them into a new resource group.  They seem to be working now, but it was a painstaking process which could have caused a major disruption as we have a mobile app and web app depending on those apps.

**Comment by sjkp** [comment]
The most staggering thing about this whole story, is that people are literally asking MS can I please pay you more to host my service, just give me an easy way to migrate. But no, we are still stuck due to the very unfortunate infrastructure stamp design of the original azure web sites (no blame on them, that was the early days of cloud - I just wish it was fixed years ago, it is not exactly a new issue, azure web sites are also plagued by it, I have sites I want to give more compute, but cannot because the stamp doesn't have any new hardware). 

One thing that is sad to read, however is that it sounds that you still are working with a inflexible stamp design (although larger stamps), but I'm fearing that we in two/three years time, will be in a similar situation, when we want to not be hosted on your old Intel hardware, and you hopefully provide more energy efficient ARM or AMD based web servers. Would it be possible to request a refresh of Yochay Kiriaty's very good article explain app service (not sure how up-to-date it is anymore) https://docs.microsoft.com/en-us/archive/msdn-magazine/2017/february/azure-inside-the-azure-app-service-architecture - a big decision today when picking between PaaS and competitive technology like Kubernetes is understanding how the PaaS offering can evolve over time.

**Comment by CRuppert** [comment]
This is a bummer that is affecting us heavily as well. Asking us to rebuild entire resource groups is just not realistic.

**Comment by wjdavis5** [comment]
This issue causes production downtime to resolve in some cases. Any updates on resolution?

**Comment by KingJohn1095** [comment]
I don't see anyone in here mentioning moving resources to a new resource group. Is this a potential workaround? 

1. Create new resource group.
2. Deploy Function App
3. Move old resources to new group.

**Comment by jeffhollan** [comment]
ACK - that is a workaround, and also updating this is still making progress.  Things slowed down a bit with COVID-19, but was just checking on progress yesterday and hoping to start another set of stamp upgrades after end of next week.

**Comment by jeffhollan** [comment]
Circling back - progress just completed.  These resource group limitations should be lifted now.

**Comment by gaikovoi** [comment]
@jeffhollan We are still facing this issue in North Central US. If I try to create new function app plan in the existing RG, I get errors:

1. AzCLI: Requested feature is not available in resource group _mygroup_. Please try using a different resource group or create a new one.
2. Portal: "code":"ValidationForResourceFailed","message":"Validation failed for a resource. Check 'Error.Details[0]' for more information.","details":[{"code":"**DSeriesWorkerNotAllowedForResourceGroup**","message":"The pricing tier PremiumV2 is not available for this resource group. Please redeploy or clone your app into a new app service plan in a new resource group. Use this link to learn more https://go.microsoft.com/fwlink/?linkid=856301"
3. REST API: The pricing tier 'ElasticPremium' is not allowed in this resource group. "ExtendedCode": "04114"

**Comment by a30003939** [comment]
Whomever fixed this on Australia SouthEast, thank you! I was about to start planning a whole database cluster migration....

**Comment by TheRealWaldo** [comment]
Still seeing this in Canada Central.   Any update?  Is this a slow roll-out?

**Comment by jeffhollan** [comment]
@TheRealWaldo I learned the other week while we fixed the initial issue, there was a platform upgrade that happened in some regions where this could pop up again (where you would need to create a new RG as the underlying stamp doesn't support premium). We have a plan to address it, the impact should be relatively small in terms of which functions fall in this category, but ETA for that is around Nov where this ideally will work flawlessly in all places.  Apologies for the issue - as soon as we got reports this was springing up we set out on a plan.

**Comment by ronnie-webb** [comment]
Still experiencing this in south central us.

**Comment by vpaiv** [comment]
We are experiencing this in West US 2.

**Comment by BMichaelJ** [comment]
Also having this issue in "Central US"
"code":"SkuNotAllowedForResourceGroup","message":"The pricing tier 'ElasticPremium' is not available for resource group

@jeffhollan  Anything new on addressing this and status?

**Comment by ronkot** [comment]
I'm still experiencing this in on of our resource groups in West Europe region while updating from `Dynamic` tier to `ElasticPremium` :

```
Upgrading from Consumption to ElasticPremium is not supported in this resource group. Please create a new ElasticPremium app directly.
```

I managed to run the similar update on two other resource groups in West Europe region, this problem occurs only on one single resource group (which happens to be the most important of these three :/).

**Comment by milanm** [comment]
Still the same in the Central US region:

> Requested feature is not available in resource group {0}. Please try using a different resource group or create a new one.

**Comment by TheRealWaldo** [comment]
Still an issue.  Destroying and rebuilding not an option.  Execs are upset with our decision to go with Azure over the competition who makes this so easy.

@jeffhollan How do we light a fire under this?

**Comment by gpetersOB** [comment]
Same issue

**Comment by bwczhao** [comment]
Now it is March of 2021, and we are having this issue in Central US. Do you have a status, or when the fix is coming?

**Comment by gitrepo12345** [comment]
The issue still persists in South Central US.

**Comment by msjonker** [comment]
@jeffhollan Any updates on this? We are experiencing this in Central US.

**Comment by anirudhgarg** [comment]
> @jeffhollan Any updates on this? We are experiencing this in Central US.

msjonker - can you confirm if yours is a Windows or a Linux Premium app.

**Comment by msjonker** [comment]
I see that this issue is related to functions, but I'm experiencing it with a Linux App Service plan. Does the fix happen to fix Resource Group dependency on webspaces/stamps as a whole?

**Comment by anirudhgarg** [comment]
@msjonker this issue was specifically for Functions. Thanks for confirming that this is not an issue with Functions. I will follow up about Web Apps.

**Comment by anirudhgarg** [comment]
> The issue still persists in South Central US.

@gitrepo12345  - can you confirm if this is for Function App and for a Windows App

**Comment by KimLClark** [comment]
Still seeing this problem when trying to add premium plan for Linux Function Apps in USWest2.  Status please?

**Comment by lecharcutier** [comment]
hey, having the same bug in a RG containing west europe consumption plan.
i have the error where trying ot deploy a premium plan "pricing tier is not available ...."
But, i can create the service plan when changing the location to west US for example

Any news on a fix or a workaround ?

**Comment by KimLClark** [comment]
Hi. Here is the workaround that has worked well for us. First, the reason for the error is that your resource group "stamp" (hardware configuration) is a version that does not support the hardware required for the Premium plan. This will not change. The simple workaround, that we are happy with, is to create a new resource group just for holding premium app service plans.  Then point your function apps there.  This configuration worked out way better than we expected. Why? Because premium app service plans are expensive. In our case we have over 15 function apps each with many functions.  You will need way less premium app service plans than you needed consumption plans (and way less servers to. Our experience is that power is about 1 premium server EP1 = 10 consumption plan servers). Having the Premium app service plans in a separate resource group allowed us to easily share one plan across a set of function apps.  Some of our function app instances have such high volumes they needed their own Premium app service plan, but for others we can share one plan across 5 function apps.  Putting the Premium plans into their own RG gives us a lot of flexibility in allocating function apps to Premium plans.

**Comment by wjdavis5** [comment]
@jeffhollan currently in a sev a and they want us to upgrade our instance to premium and then back down to S1. But we cannot b/c of this bug.

**Comment by ChrisOlivierCubed** [comment]
Thanks @KimLClark for describing your workaround, 

It seems to be the least impactful way to move only the AppPlans into a separate resource group and it may be the way which we will go as well for now, especially if you have a lot of AppPlans/Functions, redeploying all of that infrastructure to new resource groups is a horrible experience, and also, what is to be done about naming conventions, since the new resource groups will need new names, and if we want to keep with our old names, we need to do all of this twice, this is an obscene amount of effort and can easily introduce issues with 100s of functions, 

Moving the AppPlan to a newly created Resource Group also does not seem to fix this, so you have to redeploy which is not ideal

It is quite sad though, that this is still an issue 2 years after the issue was raised, (for us it is NorthEurope which still seems to have this issue), to be honest, when we already need to justify the move to Azure and now we tell business that we will need how many manhours of work to simply be able to make use of P3V3, it makes for very uncomfortable conversations

@jeffhollan Is there any update on when this is set to be completed for North Europe? Or a way in which we can initiate rollout of the fix?

**Comment by nzthiago** [comment]
Last month we identified and addressed remaining stamps where the Premium plan wasn’t enabled yet, so this will mitigate this issue further. There is only one more possibility of this issue surfacing if you already have an older Linux App Service Plan and try to create a new Linux Functions Premium plan that lands on the same scale unit. If you run into this again please raise a support ticket.
